Source/Core/Core/NetworkCaptureLogger.cpp

```diff
@@ -0,0 +1,390 @@
+  const std::string mac_address = SConfig::GetInstance().m_WirelessMac;
+  if (!Common::StringToMacAddress(mac_address, mac.data()))
+    mac.fill(0);
+
+  // Write Dolphin MAC address, the other one is dummy
+  if (is_sender)
+  {
+    // Source MAC address
+    std::copy(std::begin(mac), std::end(mac), ethernet_header.begin() + 6);
+  }
+  else
+  {
+    // Destination MAC address
+    std::copy(std::begin(mac), std::end(mac), ethernet_header.begin());
+  }
+
+  // Write the proctol type (IP = 08 00)
+  ethernet_header[12] = 8;
+  return ethernet_header;
+}
+
+std::vector<u8> PCAPSSLCaptureLogger::GetIPHeader(bool is_sender, std::size_t data_size,
+                                                  u8 protocol, s32 socket)
+{
+  constexpr std::size_t IP_HEADER_SIZE = 20;
+  std::vector<u8> ip_header(IP_HEADER_SIZE);
+  const std::size_t total_length = IP_HEADER_SIZE + data_size;
+
+  // Write IP version and header length
+  ip_header[0] = 0x45;
+  // Write total length
+  ip_header[2] = total_length >> 8 & 0xFF;
+  ip_header[3] = total_length & 0xFF;
+  // Write flags + fragment offset
+  ip_header[6] = 0x40;
+  // Write TTL
+  ip_header[8] = 0x40;
+  // Write protocol
+  ip_header[9] = protocol;
+
+  // Get socket IP address
+  struct sockaddr_in addr;
+  socklen_t addr_len = sizeof(struct sockaddr_in);
+  getpeername(socket, reinterpret_cast<struct sockaddr*>(&addr), &addr_len);
+
+  if (is_sender)
+  {
+    // Write Source IP address: 127.0.0.1
+    ip_header[12] = 0x7F;
+    ip_header[15] = 0x01;
+    // Write Destination IP address
+    std::memcpy(&ip_header[16], &addr.sin_addr, 4);
+  }
+  else
+  {
+    // Write Source IP address
+    std::memcpy(&ip_header[12], &addr.sin_addr, 4);
+    // Write Destination IP address: 127.0.0.1
+    ip_header[16] = 0x7F;
+    ip_header[19] = 0x01;
+  }
+
+  // Compute checksum
+  std::size_t index = 0;
+  u32 checksum = 0;
+  u16 word = 0;
+  for (u8 b : ip_header)
+  {
+    if (index % 2 == 0)
+      word = b;
+    else
+      checksum += (word << 8) + b;
+    index += 1;
+  }
+  checksum = (checksum + (checksum >> 16) & 0xFFFF) ^ 0xFFFF;
+
+  // Write checksum
+  ip_header[10] = checksum >> 8 & 0xFF;
+  ip_header[11] = checksum & 0xFF;
+
+  return ip_header;
+}
+
+void PCAPSSLCaptureLogger::LogTCP(bool is_sender, const void* data, std::size_t length, s32 socket)
+{
+  constexpr std::size_t TCP_HEADER_SIZE = 20;
+  std::vector<u8> tcp_header(TCP_HEADER_SIZE);
+
+  // Get socket port
+  struct sockaddr_in sock_addr;
+  struct sockaddr_in peer_addr;
+  socklen_t addr_len = sizeof(struct sockaddr_in);
+  getsockname(socket, reinterpret_cast<struct sockaddr*>(&sock_addr), &addr_len);
+  getpeername(socket, reinterpret_cast<struct sockaddr*>(&peer_addr), &addr_len);
+
+  // Write port
+  if (is_sender)
+  {
+    std::memcpy(&tcp_header[0], &sock_addr.sin_port, 2);
+    std::memcpy(&tcp_header[2], &peer_addr.sin_port, 2);
+
+    // Write sequence number
+    static u32 sequence_number = 0;
```

Comment:
Is there any reason this and the other function-local mutable statics aren't class variables? This makes the class thread-unsafe in its current form.